### ISSUE TITLE
[Feature Requests] - Add selection tools, and the ability to work with/on selections

### ISSUE THREAD

**Issue by achadwick** [mypaint/mypaint#31]
Migrated here from [Gna! bug 19207](https://gna.org/bugs/index.php?19207) because this is such a frequently requested feature.

> Okay, this is a slightly obvious, yet slightly important one: 
> Sometimes, an artist may wish to simply move something they had drawn, either to make room for something else, or to reposition something mispositioned. 
> This cannot be done unless there are tools to create a selection (either through circles, squares, freeform, etc.), and cut/copy/paste/move/otherwise transform it (or the entire layer's contents) 
> As long as its not as overly complicated as The GIMP's excuse for this feature (which requires going through 3 different tools to move/resize/etc.), this would be great.
> -- [. <lirodon>](https://gna.org/users/lirodon)

No time to work on this before 1.2, deferring it again. Some mockups would be nice: it's fairly obvious how selection creation should work, but how to work _with_ selections? What should the user be able to do with them?
- Transform to a layer (& thus to a mask?)
- Accept brushstroke dabs only within the selection
- Cut/copy/paste
- Er, what else?

**Comment by Shikei** [comment]
I'm dying for a selection tool. even just the free-hand selection with a transform tool

**Comment by ayberkozgur** [comment]
Lack of cut/paste into different layer is in my opinion the number one shortcoming of mypaint. Even a very simple rectangular cut&paste would improve usage drastically.

**Comment by eduardomezencio** [comment]
I see lots of people that seem to need some 'advanced' functionality and, of course, all of them can really be useful, but I think just having a rectangular and maybe a free-hand selection tools with cut, paste and move would make so much difference to basically every user of mypaint right now that maybe we're better off with just that for now.
Of course I may be wrong, but I think if developers would implement all transform-related functions and this kind of stuff, 20% of the effort would go to making stuff that would be used 80% of the time (copy, paste, move) and 80% of the effort to make stuff that would be used 20% of the time. Of course I'm making those numbers up, just to illustrate my concern.
I will never be against adding cool stuff to the program, please, don't get me wrong. I'm just spitting some thoughts

**Comment by achadwick** [comment]
@freelanceear writes in #443:
> Rotating layers would be a very useful feature for adjusting sketches pre-paint. Moving layers is implemented, as is rotating the canvas view, so it just seems logical to be able to rotate layers. As for the implementation, I think holding Ctrl while moving a layer to rotate would work nicely.

**Comment by nicolas33** [comment]
I tend to be regularly fustrated when I realize I have to move a region and just can't do it in my drawing software. MyPaint is one of the BEST drawing soft I've used but the lack of this feature make the adventure far from fun all the time.

Please, rank this with a higher priority! As a user, I don't care how it's implemented. Even a beginning of something could be very helpfull. Mark the feature as TESTING, EXPERIMENTAL, WILL_CRASH_EVERYTHING or whatever. But please, release something in this area! Please, please, please !

**Comment by achadwick** [comment]
@nicolas33 I figure that's a reasonable request for the first bit of work after the 1.2.0 release is out. Reassigned.

**Comment by earcanal** [comment]
+1

**Comment by butu5** [comment]
+1
Lasso Selection, Ability to Copy, Paste & transform is really needed in MyPaint. Only because of this I am unable to make MyPaint as my primary tool :(

**Comment by achadwick** [comment]
@butu5 Sure! I really want to help bring new contributors up to speed. I guess the right place to ask is in the new Community forums - we have a developer's area for just this sort of semi-mentoring. Come and say hi over there!

→ <http://community.mypaint.org/c/development>

I've been expanding the wiki's “how to get involved” pages recently too:

→ <https://github.com/mypaint/mypaint/wiki/Contributing>

and I'll try to drop a few more notes in there today, though bear in mind that I'm travelling.

**Comment by feeef** [comment]
+1 for this feature that MyPaint must have!

I have programming skills and I am willing to help. Have you started implementing the feature? How can I help with the code?

Thanks a lot for this marvelous software anyway!

**Comment by OrenjiAkira** [comment]
how is this issue going? any plans towards future versions?

**Comment by achadwick** [comment]
Is anyone else working on this? I have a few middle-distance plans.

@OrenjiAkira 
A few more bits of groundwork needed first, but I'm quietly working on it. First one is a small refactor to allow our "Brushwork" code to capture your brushstrokes into a layer other than the current one. I want to try that bit out by implementing a clone tool first. Clone tools can work by using your brush to make a temporary mask layer which gets stamped down along with an offset pattern when you lift your brush up. So it's a bit different, but it's work that relies on an out-of-tree mask layer, so it's probably good practice :sweat_smile:

Selection-era Cut is the application of a Destination-Out op with a mask to the source layer(s), and both Copy and Cut can use a Destination-In with the mask to get the cutout itself. Beyond that I'm not sure about the UI except that Paste is probably where we'll hang some sort of transformation cage, meaning it should probably be its own temporary mode. The minimal Cairo stuff in commit d6e76c98cfad1213d84820e43386789a8f783a3e will be able to do affine transformations (translate+scale+rotate+skew) but nothing more mathematically complex (like perspective or cage deforms).

(Speculative stuff now...)

During a paste in this model, each transform would require a re-rendering of the pasted pixbuf with Cairo. This could be a bit slow for a huge paste layer (MyPaint layers routinely get huge), but faster code for simply offsetting the current preview exists in the Move Layer tool.

(Big question too: how does pasting/transforming work with the undo stack? Do users get to undo and redo all stages of the paste/transform?)

This side all sounds hard. Practicality-wise, the first kind of selection-era Paste operation I implement will probably just make a new layer and maybe drop you into Move Layer mode of that feels comfortable.

**Comment by odysseywestra** [comment]
Link backing the thread on our community fourms: https://community.mypaint.org/t/add-selection-tools/132

---
Tirage Note: Highest requested feature on both sites.

**Comment by dandv** [comment]
Also the first feature I looked for when I launched the app. 🤦

**Comment by odraencoded** [comment]
I have no idea how this is taking over 6 years. There's literally only three essential features:

* Select rectangle.
* Fill selection. (delete = transparent color)
* Copy selection. (cut = copy + delete)

If **just** this was added everyone could move parts of a sketch without having to switch programs simply by cutting, pasting, moving, and merging. It should also be trivial to add mirror/rotate 90 degrees to layers, which is currently missing from MyPaint.

Even without free transform and scaling, and without making it so brushes only paint the selected area, this extremely simple selection tool would improve MyPaint greatly. Without the ability to reposition parts of layers MyPaint is currently inferior to MSPaint. I'm not even sure there's any other image editing or drawing program in the world that lacks this function, so honestly even if you call it 2.0 it's in practice currently an incomplete software.

Right now I bet a lot of people download MyPaint and think "wow this canvas, these brushes, it's all amazing," and then they need move a character's head a few pixels to the side, google how to, try to use the edit layer menu and it opens in MSPaint without alpha instead, or they try to copy and paste it in photoshop and it pastes a fully black image, or finally download GIMP just to paste the sketch and it's all black pencil with a checkered alpha background and you can't even see what you're moving but it works but it's such a hassle to do something so basic that they just give up and go use Krita or Fire Alpaca instead because at least they have selection tools.

I'm not even sure it's possible to make image-editing selection work perfectly in MyPaint, given that MyPaint has stuff like mirrors. Would moving a selection in a mirror layer move the other side too? Who would even use this? Why was mirroring implemented before the ability to mirror layers horizontally??? I agree it's nice but these priorities literally couldn't make any less sense.

Some thoughts:

* Rectangle and lasso is probably enough, Ellipsis, etc.  are probably not useful in a drawing program, they're good in image editing software when you work with the mouse, not the tablet. Maybe magic wand would be nice too, using the same code as bucket fill?
* The select tool and the transform tool should be separate. Selecting something doesn't automatically switch to transforming.
* Using a select tool while a selection already exists replaces the selection entirely. Shift/alt/ctrl modifiers let users add or remove parts to the selection mask. If you select a rectangle the wrong size, it's intuitive to just select a new rectangle the right size. Absolutely **NO HANDLES** to resize the selection. **NO DRAGGING** the selected area around either. You click with the selection tool ANYWHERE on the canvas, the previous selected area goes poof. It's that simple and intuitive. Interactive selection areas ruin with this concept, so the selection area should be purely visual when the select tool is active.
* Some software like Paint Tool SAI lets you use a brush-like selection tool to paint/erase the selection mask, which is shown blue.
* Transform tools in a drawing program are used to correct placement mistakes, they aren't used to precisely move around objects. You don't need every transform change to be part of the undo-stack. You want a transform editing mode that's temporary and starts once you click on a transform tool, then on the transform handles, and that lets you edit the transform, and then when you apply the transform the whole transform is committed at once.
* If the user transforms the selection wrong and undoes, the previous point in history should have the selection mask intact so they can quickly transform again. (in other words: if you undo and the selection disappears, that sucks because you have to select everything again). Photoshop, for example, achieves this by adding every change to the selection mask to the history.
* Select all isn't necessary. Commands like "clear layer" "copy layer" should just be changed to "clear" and "copy." If there's a selection, it affects the selection, otherwise it affects the whole layer. The same thing applies to transform tools: without a selection, it transforms the whole layer.
* Invert selection should be limited to the area of the selected layer, not the whole image, but you should be able to select areas beyond the current layer's boundaries, so that things like "copy merged" can work. Using copy merged without a selection would copy the whole image (clipped by the frame). Inverting the selection without anything selected = Select All (i.e. select the whole layer area).
* There could be a button to quickly create a mask (e.g. DestinationIn layer) straight from a selection, preferably wrapping the select layer in a group. Similarly, a button to set the image frame (i.e. clip) to the boundaries of the selection.
* You should be able to transform a whole at once, even if the group doesn't have pixels for you to select.

All this very useful stuff for some reason has nothing to do with limiting the brush strokes to the selection area. I'm pretty sure nobody even cares about that feature. Like, it's a drawing program. The brushes are all soft and textured, but the edges of a rectangle or lasso are sharp, so if you drew a stroke at the edge of the selection, it would look jarred and ugly. Why would anyone do this? It's prettier to use a new layer and an eraser, or use a mask. Does anybody use selection areas to draw with brushes on it? Maybe if you wanted to fill the whole thing with a single color, or quickly add a texture, but both things can be done with masks, so I guess this is actually the least important selection feature and probably the one that would take most time to implement.

**Comment by jplloyd** [comment]
@odraencoded
Sounds good, I agree with pretty much all of your points. When you implement this, don't forget about handling the stroke maps as well.

As for the external editing, it works just fine on linux. Unfortunately no windows-using devs have volunteered their time to improve matters for that OS.

> I have no idea how this is taking over 6 years.

Turns out these things don't build themselves, no matter how long you wait. Bear in mind that nobody gets paid for working on this project. 

You yourself have contributed 1 translated string and (effectively) 1 line of code to this repository and I'm sure you could implement this in another 2-3k lines. If you implement this such that the new code is as insulated as possible, I will definitely review/merge it. If not, you will have to wait until somebody else does. Complaining is unlikely to speed the process up and might just have the adverse effect.

> even if you call it 2.0 it's in practice currently an incomplete software.

The reason it's 2.0 is because the .ora files are not strictly backwards compatible with 1.2.1 and earlier versions, due to both parameters stored in the stroke maps and the switch to linear sRGB. It has nothing to do with any notion of feature-completeness.

**Comment by odraencoded** [comment]
@jplloyd 
I don't understand the point of having "feature requests" if you're going to say "if you want that done just do it yourself."

>When you implement this, don't forget about handling the stroke maps as well.

???

In case I didn't make my point clear enough:

>Sometimes, an artist may wish to simply **MOVE** something they had drawn, either to make room for something else, or to **REPOSITION SOMETHING MISPOSITIONED**.
>This cannot be done unless there are tools to create a selection (either through circles, squares, freeform, etc.), and **CUT/COPY/PASTE/MOVE/OTHERWISE TRANSFORM** it (or the entire layer's contents)

>I'm dying for a selection tool. even just the free-hand selection with **A TRANSFORM TOOL**.
that would make mypaint definitely the best painting software of all

>Lack of **CUT/PASTE INTO DIFFERENT LAYER** is in my opinion the number one shortcoming of mypaint. **EVEN A VERY SIMPLE RECTANGULAR CUT&PASTE** would improve usage drastically.

>I see lots of people that seem to need some 'advanced' functionality and, of course, all of them can really be useful, but I think just having a rectangular and maybe a free-hand selection tools with **CUT, PASTE AND MOVE** would make so much difference to basically every user of mypaint right now that maybe we're better off with just that for now.

>I tend to be regularly fustrated when I realize I have to **MOVE** a region and just can't do it in my drawing software. MyPaint is one of the BEST drawing soft I've used but the lack of this feature make the adventure far from fun all the time.

>+1 for rectangular **SELECT, MOVE, CUT, COPY, PASTE** as a MVP.

>Lasso Selection, Ability to **COPY, PASTE & TRANSFORM** is really needed in MyPaint. Only because of this I am unable to make MyPaint as my primary tool :(

>By selection tools I mean example a lasso tool, so **I CAN SELECT, CUT AND PASTE** a portion of my drawing, magic wand etc…

>This is the feature that is most needed to make it as useful as any other paint program. The ability **TO SELECT A REGION, MOVE, ROTATE AND SCALE IT**

>Adding my vote for **SELECTION/MANIPULATION** tools

>You can also duplicate the layer, erase everything on the duplicate that you don’t want to move, and then **MOVE** the layer.

>Its not about selection its about **MOVE PART OF THE PAINTING**.

>I am still waiting for **SELECTION - TRANSFORMATION** tool

>a selection and even some **BASIC TRANSFORM TOOLS** will really go a long way.

There isn't a single user saying they want to draw in the selection. Nobody. Not even one. The only time selection masking is mentioned is when comparing with selection tools of other image editing software where the selection tools work that way. That sort of selection isn't what MyPaint users have been begging for for six years.

You could literally make a lasso tool that just cuts part of the current layer and creates a new layer with that part, plus transform tools, and voilá everybody rejoices. No need to even touch stroke-related code.

**Comment by nekohayo** [comment]
@odraencoded your tone is abusive, and I would highly recommend you calm down and stop the vitriol *right now.* I'm just a fellow user but I feel bad for the maintainers on the receiving end of your rant. If I was an admin, I'd ban you outright after that second comment. These days I have no patience for disrespect.

Either you contribute code (because clearly it sounds "so easy" when you say it, which is why I suppose @jplloyd challenged you to produce a patch), or bring in someone who can, or stop complaining aggressively about an application you are getting _for free_ (built from other people's volunteer time, no less!)

**Comment by Marand-** [comment]
Wow, I'm surprised this is still an open issue and I'm getting notifications on it.  It'a amazing this is still an active issue, considering I commented here over six years ago, and the original discussion on the now-defunct gna.org was even older than that.  @odraencoded is right that 6+ years for even basic selection tools is crazy; Krita has managed to reimplement most of MyPaint's useful features in less time.  

> If I was an admin, I'd ban you outright after that second comment.

Anyway, @nekohayo, if you're going to be tone-policing discussions here you should probably start with @jplloyd, considering that's where things went downhill from what I see.  The original comment seemed fine to me until @jplloyd's snarky, passive-aggressive response.  If this had been said to me I wouldn't have been happy either:

> When you implement this, don't forget about handling the stroke maps as well.

> You yourself have contributed 1 translated string and (effectively) 1 line of code to this repository and I'm sure you could implement this in another 2-3k lines.

That, plus the surrounding context, gives the entire comment a passive-aggressive tone of "we don't care, do it yourself.  Oh you won't, because you've only contributed one translated string, which barely even counts as contribution, LOL".  

Adding that "[c]omplaining is unlikely to speed the process up and might just have the adverse effect" is also petty considering the original comment was overall reasonable and attempted to add useful input as well as ask what was going on that a feature considered basic for these types of tools has been sitting in development hell for 6+ years.

He(?) obviously cared enough about the software to take the time to write a lengthy comment providing input, and a response like that is insulting and discouraging.  It's no surprise that the tone in the followup response degraded like it did.

And I'll get this out of the way before someone comments on my lack of contribution as well:  I used MyPaint quite regularly in the past and used to be pretty active with bug reports, suggestions, and trying to help other users.  I haven't been in a while because Krita consumed most of my MyPaint use cases, though, so I was surprised to come back to this discussion and see one of the maintainers responding like a petulant child over some mild criticism.

**Comment by odysseywestra** [comment]
I'm going to go ahead and close this issue since the conversation isn't really helping. When someone has solution to address adding this feature, feel free to open a pull request and we'll help you get into the code base. For the rest of the community, feel free to help build the scope of this feature request on our community forums. https://community.mypaint.org/t/add-selection-tools/132 Thanks.  

PS. Please treat each other with respect. I know this is highly requested feature, but please don't try to scare contributors away. That's how projects die.

**Comment by jplloyd** [comment]
@odraencoded 
I'm sorry for being snide in my comment, it was completely unnecessary. Particularly bringing up your contributions was really stupid, since any contribution really is appreciated. The point I was trying to make, and should have made in a better way, was that even a simple implementation requires a certain amount of work and someone has to actually do that work.

I definitely overreacted and should have just commented on them separately, but I hope you also agree that remarks like these in your original comment were not really helpful or necessary.

> Without the ability to reposition parts of layers MyPaint is currently inferior to MSPaint.

> ...honestly even if you call it 2.0 it's in practice currently an incomplete software.

> ...these priorities literally couldn't make any less sense.

Ignoring some adjectives (as in "extremely simple") and remarks like those, your input is 100% helpful.

I should have clarified what stroke maps are, because they are not related to masking at all. Stroke maps are what allow you to restore the brush (and optionally color) by clicking on a key/button while hovering over a stroke in a layer. The transformations of selected areas would have to also deal with these stroke maps (or toss them away for the entire layer or layers) or the visuals would no longer match the stroke data.

This is just an example of a complication that you might not have known about. I don't think it's a huge amount of extra work, but just another thing that has to be taken into account.

My point about time still stands. It makes no difference whether we wait one year or 10 years for a thing if nobody is working on it.

I'm going to reopen this, since it _is_ a widely requested and arguably essential feature. Again I apologize for derailing the discussion and for being a snide bastard.

**Comment by feeef** [comment]
If this feature is as requested as it seems there might be a way to finance its implementation through a platform like bountysource or through another way ?

**Comment by jplloyd** [comment]
@feeef Certainly a possibility. There used to be a bountysource account for the project, but it appears to have been de-registered (I'm also not sure if anyone active in the project had access to it). 

The only problem I can think of would be if multiple people started working on it simultaneously, but I assume there are some methods in place to avoid that (probably quite unlikely anyway).

It would also require a detailed scope and set of requirements, so there's as little room for interpretation as possible. Those could either be placed in the first post of this issue, or in a new issue.

**Comment by stevphie** [comment]
Just installed mypaint, looking for selection tools and couldn't find it. I thought I might be missing something or what... but then I stumbled upon lasso feature request from 2014 which is not solved yet.

Pardon... But, holy mother of god

**Comment by PJSingh5000** [comment]
Surprised that a rectangular select feature is not present.

**Comment by EmilPi** [comment]
> Wow, I'm surprised this is still an open issue and I'm getting notifications on it. It'a amazing this is still an active issue, considering I commented here over six years ago, and the original discussion on the now-defunct gna.org was even older than that. @odraencoded is right that 6+ years for even basic selection tools is crazy; Krita has managed to reimplement most of MyPaint's useful features in less time.
> 
> > If I was an admin, I'd ban you outright after that second comment.
> 
> Anyway, @nekohayo, if you're going to be tone-policing discussions here you should probably start with @jplloyd, considering that's where things went downhill from what I see. The original comment seemed fine to me until @jplloyd's snarky, passive-aggressive response. If this had been said to me I wouldn't have been happy either:
> 
> > When you implement this, don't forget about handling the stroke maps as well.
> 
> > You yourself have contributed 1 translated string and (effectively) 1 line of code to this repository and I'm sure you could implement this in another 2-3k lines.
> 
> That, plus the surrounding context, gives the entire comment a passive-aggressive tone of "we don't care, do it yourself. Oh you won't, because you've only contributed one translated string, which barely even counts as contribution, LOL".
> 
> Adding that "[c]omplaining is unlikely to speed the process up and might just have the adverse effect" is also petty considering the original comment was overall reasonable and attempted to add useful input as well as ask what was going on that a feature considered basic for these types of tools has been sitting in development hell for 6+ years.
> 
> He(?) obviously cared enough about the software to take the time to write a lengthy comment providing input, and a response like that is insulting and discouraging. It's no surprise that the tone in the followup response degraded like it did.
> 
> And I'll get this out of the way before someone comments on my lack of contribution as well: I used MyPaint quite regularly in the past and used to be pretty active with bug reports, suggestions, and trying to help other users. I haven't been in a while because Krita consumed most of my MyPaint use cases, though, so I was surprised to come back to this discussion and see one of the maintainers responding like a petulant child over some mild criticism.

Thank you for mentioning Krita. Moving to it right from this 6-years long feature requests.
Thanks the author for the all hard work, but in today's world there is some bar to be supported...

**Comment by jplloyd** [comment]
I think the existing comments cover everything about ideas, requirements, etc. so to avoid further noise I'm locking the topic.

I deleted a comment that basically said that [Krita](https://krita.org/en/) is a good alternative, but elected to quote a huge chunk of text in order to do so. I completely agree with the comment, but it also made me realize it's very unlikely further comments will be constructive.

**Comment by AesaraB** [comment]
If I were ever to address this, it would have to happen after #209
